packages/engine/src/scene/systems/ShadowSystem.ts

```diff
@@ -171,108 +184,86 @@ export default async function ShadowSystem(world: World) {
 
-      return function cleanup() {}
-    }, [shadowComponents, useShadows, shadowMaterial])
+      const radius = Math.max(sphere.radius * 2, minRadius)
+      const center = groupComponent.value[0].worldToLocal(sphere.center)
+      const shadowEntity = createEntity()
+      const shadowObject = new Mesh(shadowGeometry, shadowMaterial.value.clone())
+      addObjectToGroup(shadowEntity, shadowObject)
+      addComponent(shadowEntity, NameComponent, 'Shadow for ' + getComponent(entity, NameComponent))
+      addComponent(shadowEntity, VisibleComponent)
+      setComponent(entity, DropShadowComponent, { radius, center, entity: shadowEntity })
+
+      return () => {
+        removeComponent(entity, DropShadowComponent)
+        removeEntity(shadowEntity)
+      }
+    }, [useShadows, shadowMaterial, groupComponent, shadow])
 
     return null
   })
 
   const shadowOffset = new Vector3(0, 0.01, 0)
 
   const execute = () => {
-    const setDropShadowMatrix = (matrix: Matrix4, entity: Entity, caster: Entity) => {
-      const transformComponent = getComponent(entity, TransformComponent)
-      transformComponent.matrix.copy(matrix)
-    }
-
-    sceneObjects = Array.from(Engine.instance.currentWorld.objectLayerList[ObjectLayers.Camera] || [])
+    sceneObjects = Array.from(Engine.instance.currentWorld.objectLayerList[ObjectLayers.Scene] || [])
 
     const useShadows = getShadowsEnabled()
     if (!useShadows) {
       for (const entity of dropShadowComponentQuery()) {
-        const dropShadowComponent = getComponent(entity, DropShadowComponent)
-
-        if (!entityExists(dropShadowComponent.entity)) {
-          removeComponent(entity, DropShadowComponent)
-          continue
-        }
+        const dropShadow = getComponent(entity, DropShadowComponent)
+        const dropShadowTransform = getComponent(dropShadow.entity, TransformComponent)
 
         raycaster.firstHitOnly = true
-        raycasterPosition.copy(dropShadowComponent.center)
+        raycasterPosition.copy(dropShadow.center)
         getComponent(entity, GroupComponent)[0].localToWorld(raycasterPosition)
         raycaster.set(raycasterPosition, shadowDirection)
 
         const intersected = raycaster.intersectObjects(sceneObjects)[0]
         if (!intersected || !intersected.face) {
-          setDropShadowMatrix(defaultShadowMatrix, dropShadowComponent.entity, entity)
+          dropShadowTransform.scale.setScalar(0)
           continue
         }
 
-        const centerCorrectedDist = Math.max(intersected.distance - dropShadowComponent.center.y, 0.0001)
+        const centerCorrectedDist = Math.max(intersected.distance - dropShadow.center.y, 0.0001)
 
         //arbitrary bias to make it a bit smaller
         const sizeBias = 0.3
-        const finalRadius =
-          sizeBias * dropShadowComponent.radius + dropShadowComponent.radius * centerCorrectedDist * 0.5
+        const finalRadius = sizeBias * dropShadow.radius + dropShadow.radius * centerCorrectedDist * 0.5
```

Comment:
Super minor suggestion, but would it be helpful to use math.min to cap the final radius to some value, maybe 2*dropShadow.radius? This would prevent shadow planes from becoming enormous as distance increases.